### PR TITLE
50 in sql statements values are inserted as parameters with placeholders

### PR DESCRIPTION
**Pull request by DrDiHai**

### CODE REVIEW

File: backend/src/database/sqlexecutable.py [hunk 1]
```diff
@@ -182,13 +203,28 @@ def __init__(
         self._script = (
             script_or_template
             if isinstance(script_or_template, str)
-            else self.__class__.sql_templates.get(script_or_template).format(**kwargs)
+            else self.__class__.sql_templates.get(script_or_template)
         )
+        self._register_and_replace_named_parameters(self._script, kwargs)
```
Comment (hk-ret) [inline backend/src/database/sqlexecutable.py:208]: see comment below
```suggestion
        self._script = self._register_and_replace_named_parameters(self._script, kwargs)
```

File: backend/testing/database/test_sqlite.py [hunk 2]
```diff
@@ -50,131 +51,104 @@ def setUp(self) -> None:
...
         mock_con_obj.connect = AsyncMock(return_value=mock_con_obj)
         Mock_Connection = Mock(return_value=mock_con_obj)
-        with (patch("db.sqlite.SQLiteConnection", Mock_Connection),):
+        with (patch("database.sqlite.SQLiteConnection", Mock_Connection),):
             reply = await self.db.connect()
             self.assertEqual(reply, mock_con_obj)
             Mock_Connection.assert_called_once_with(db_obj=self.db, **self.db_cfg)
             mock_con_obj.connect.assert_awaited_once_with()
 
     def test_201_sql_table_list(self):
```
Comment (hk-ret) [inline backend/testing/database/test_sqlite.py:91]:
```suggestion
    @unittest.skip('to be adapted when sql_factory is redesigned')
    def test_201_sql_table_list(self):
```

File: backend/testing/database/test_sqlite.py [hunk 3]
```diff
@@ -50,131 +51,104 @@ def setUp(self) -> None:
...
-        reply = self.db.sql(db.sql.SQL.TABLE_LIST)
-        re = "^ *SELECT.*FROM sqlite_master.*'table'"
+        reply = self.sql.script(database.sqlexecutable.SQLTemplate.TABLELIST).sql()
+        re = "^ *SELECT name .*FROM sqlite_master.*'table'"
         re += ".*substr.*'sqlite_' *$"
         self.assertRegex(reply.replace("\n", " "), re)
 
-    def test_202_sql_table_info(self):
+    async def test_202_get_table_info(self):
```
Comment (hk-ret) [inline backend/testing/database/test_sqlite.py:97]:
```suggestion
    @unittest.skip('to be adapted when sql_factory is redesigned')
    async def test_202_get_table_info(self):
```

File: backend/testing/database/test_sqlkeymanager.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        self.assertEqual(key, "param")
+        self.assertEqual(self.SQLKeyManager._keys, {"param"})
+
+    def test_register_key_empty(self):
+        key = self.SQLKeyManager.register_key("")
+        self.assertEqual(key, "1")
+        self.assertEqual(self.SQLKeyManager._keys, {"1"})
+
+    def test_register_key_duplicate(self):
+        self.SQLKeyManager._keys = {"param"}
```
Comment (hk-ret) [inline backend/testing/database/test_sqlkeymanager.py:33]:
```suggestion
        self.assertEqual(self.SQLKeyManager.register_key("param"), "param")
```

File: backend/src/database/sqlexecutable.py (hunk 1 above)
Comment (DrDiHai) [reply backend/src/database/sqlexecutable.py:208]: Als ein commit übernommen

File: backend/src/database/sqlexecutable.py [hunk 5]
```diff
@@ -182,13 +203,28 @@ def __init__(
...
 
+    def _register_and_replace_named_parameters(
+        self, query: str, params: dict[str, SQLDataType]
+    ):
+        for key, value in params.items():
+            if not key in query:
+                raise ValueError(f"Parameter '{key}' not found in query '{query}'.")
+            final_key = self._create_param(key, value)
+            query = query.replace(f":{key}", f":{final_key}")
+
```
[the remark this answers is not in the document]
Comment (DrDiHai) [reply backend/src/database/sqlexecutable.py:222]: Als ein commit übernommen

File: backend/src/database/sqlexecutable.py [hunk 6]
```diff
@@ -204,23 +225,28 @@ def __init__(
...
         self._columns.append(
             self.sql_factory.get_sql_class(SQLColumnDefinition)(
                 name, data_type, constraint, **pars
             )
         )
         return self
 
-    def sql(self) -> str:
+    def get_params(self):
+        value_dict = {}
```
[the remark this answers is not in the document]
Comment (DrDiHai) [reply backend/src/database/sqlexecutable.py:259]: Agreed, now returns an empty dict

File: backend/src/database/sqlexpression.py [hunk 7]
```diff
@@ -200,13 +246,27 @@ def __init__(
...
 
-    def sql(self) -> str:
+    def get_params(self):
+        return {
+            **self.first.get_params(),
+            **self.second.get_params(),
+            **self.third.get_params(),
+        }
+
+    ##### Need to add support to have lists and dictionaries as values that are serialized as json #####
```
[the remark this answers is not in the document]
Comment (DrDiHai) [reply backend/src/database/sqlexpression.py:262]: Was an old todo-comment, removed it

File: backend/src/database/sqlexpression.py
[the remark this answers is not in the document]
Comment (DrDiHai) [reply backend/src/database/sqlexpression.py:1]: Not sure about a good name - SQLClause for now, but can now be easily changed again ;).

**Pull request merged**